### ISSUE TITLE
rrdesi+GPU problems with large input files

### ISSUE THREAD

**Issue by sbailey** [desihub/redrock#222]
`rrdesi_mpi --gpu ...` has problems with large input files, e.g. healpix of sv3 rosettes or areas of the main survey where we have completed all layers.

To reproduce, create an example file with 5k targets:
```python
from desispec.io import read_spectra, write_spectra
from desispec.spectra import stack

basedir = '/global/cfs/cdirs/desi/spectro/redux/himalayas/tiles/cumulative/1000/20210517'
coadds = list()
for petal in range(10):
    coadds.append(read_spectra(f'{basedir}/coadd-{petal}-1000-thru20210517.fits'))

coadd = stack(coadds)
write_spectra('coadd-5k.fits', coadd)
```
there is a copy of this in /pscratch/sd/s/sjbailey/temp/coadd-5k.fits but I don't promise to keep it there forever, so consider making your own copy.

The following examples were done on an interactive GPU node session:
```
salloc -N 1 -q interactive -C gpu -t 2:00:00 -A desi_g --gpus-per-node=4
```

Running in CPU mode works fine:
```
time srun -n 64 -c 2 --cpu-bind=cores rrdesi_mpi -i coadd-5k.fits -o redrock-5k-cpu.fits
```

But running in GPU mode crashes in various ways:
```
time srun -n 64 -c 2 --cpu-bind=cores --gpu-bind=map_gpu:3,2,1,0 \
    rrdesi_mpi -i coadd-5k.fits -o redrock-5k-gpu.fits --gpu --max-gpuprocs 4
```

On the main branch:
```
...
Scanning redshifts: 177.4 seconds
srun: error: nid001228: task 0: Bus error
srun: launch/slurm: _step_signal: Terminating StepId=3839955.1
slurmstepd: error: *** STEP 3839955.1 ON nid001228 CANCELLED AT 2022-12-02T16:20:29 ***
srun: error: nid001228: tasks 4-15,55-57,59-60,62-63: Terminated
srun: error: nid001228: tasks 16-27,29-54,58,61: Terminated
srun: error: nid001228: task 28: Terminated
srun: error: nid001228: task 3: Terminated
srun: error: nid001228: task 2: Terminated
srun: error: nid001228: task 1: Terminated
srun: Force Terminated StepId=3839955.1
```

On the gpu branch in PR #211 
```
...
--- Process 1 raised an exception ---
Proc 1: Traceback (most recent call last):
Proc 1:   File "/global/common/software/desi/users/sjbailey/redrock/py/redrock/external/desi.py", line 787, in rrdesi
    scandata, zfit = zfind(targets, dtemplates, mpprocs,
Proc 1:   File "/global/common/software/desi/users/sjbailey/redrock/py/redrock/zfind.py", line 202, in zfind
    local_targets, results = _rebalance_after_scan(targets, results)
Proc 1:   File "/global/common/software/desi/users/sjbailey/redrock/py/redrock/zfind.py", line 107, in _rebalance_after_scan
    lopsided_targets = targets.comm.gather(targets.local(), root=0)
Proc 1:   File "mpi4py/MPI/Comm.pyx", line 1578, in mpi4py.MPI.Comm.gather
Proc 1:   File "mpi4py/MPI/msgpickle.pxi", line 772, in mpi4py.MPI.PyMPI_gather
Proc 1:   File "mpi4py/MPI/msgpickle.pxi", line 147, in mpi4py.MPI.pickle_dump
Proc 1:   File "mpi4py/MPI/msgbuffer.pxi", line 50, in mpi4py.MPI.downcast
Proc 1: OverflowError: integer 2277326910 does not fit in 'int'

MPICH Notice [Rank 1] [job id 3839955.3] [Fri Dec  2 08:28:03 2022] [nid001228] - Abort(0) (rank 1 in comm 0): application called MPI_Abort(MPI_COMM_WORLD, 0) - process 1

aborting job:
application called MPI_Abort(MPI_COMM_WORLD, 0) - process 1
--- Process 2 raised an exception ---
Proc 2: Traceback (most recent call last):
Proc 2:   File "/global/common/software/desi/users/sjbailey/redrock/py/redrock/external/desi.py", line 787, in rrdesi
    scandata, zfit = zfind(targets, dtemplates, mpprocs,
Proc 2:   File "/global/common/software/desi/users/sjbailey/redrock/py/redrock/zfind.py", line 202, in zfind
    local_targets, results = _rebalance_after_scan(targets, results)
Proc 2:   File "/global/common/software/desi/users/sjbailey/redrock/py/redrock/zfind.py", line 107, in _rebalance_after_scan
    lopsided_targets = targets.comm.gather(targets.local(), root=0)
Proc 2:   File "mpi4py/MPI/Comm.pyx", line 1578, in mpi4py.MPI.Comm.gather
Proc 2:   File "mpi4py/MPI/msgpickle.pxi", line 772, in mpi4py.MPI.PyMPI_gather
Proc 2:   File "mpi4py/MPI/msgpickle.pxi", line 147, in mpi4py.MPI.pickle_dump
Proc 2:   File "mpi4py/MPI/msgbuffer.pxi", line 50, in mpi4py.MPI.downcast
Proc 2: OverflowError: integer 2279388450 does not fit in 'int'

MPICH Notice [Rank 2] [job id 3839955.3] [Fri Dec  2 08:28:04 2022] [nid001228] - Abort(0) (rank 2 in comm 0): application called MPI_Abort(MPI_COMM_WORLD, 0) - process 2

aborting job:
application called MPI_Abort(MPI_COMM_WORLD, 0) - process 2
--- Process 3 raised an exception ---
Proc 3: Traceback (most recent call last):
Proc 3:   File "/global/common/software/desi/users/sjbailey/redrock/py/redrock/external/desi.py", line 787, in rrdesi
    scandata, zfit = zfind(targets, dtemplates, mpprocs,
Proc 3:   File "/global/common/software/desi/users/sjbailey/redrock/py/redrock/zfind.py", line 202, in zfind
    local_targets, results = _rebalance_after_scan(targets, results)
Proc 3:   File "/global/common/software/desi/users/sjbailey/redrock/py/redrock/zfind.py", line 107, in _rebalance_after_scan
    lopsided_targets = targets.comm.gather(targets.local(), root=0)
Proc 3:   File "mpi4py/MPI/Comm.pyx", line 1578, in mpi4py.MPI.Comm.gather
Proc 3:   File "mpi4py/MPI/msgpickle.pxi", line 772, in mpi4py.MPI.PyMPI_gather
Proc 3:   File "mpi4py/MPI/msgpickle.pxi", line 147, in mpi4py.MPI.pickle_dump
Proc 3:   File "mpi4py/MPI/msgbuffer.pxi", line 50, in mpi4py.MPI.downcast
Proc 3: OverflowError: integer 2279025418 does not fit in 'int'

MPICH Notice [Rank 3] [job id 3839955.3] [Fri Dec  2 08:28:04 2022] [nid001228] - Abort(0) (rank 3 in comm 0): application called MPI_Abort(MPI_COMM_WORLD, 0) - process 3

aborting job:
application called MPI_Abort(MPI_COMM_WORLD, 0) - process 3
--- Process 0 raised an exception ---
Proc 0: Traceback (most recent call last):
Proc 0:   File "/global/common/software/desi/users/sjbailey/redrock/py/redrock/external/desi.py", line 787, in rrdesi
    scandata, zfit = zfind(targets, dtemplates, mpprocs,
Proc 0:   File "/global/common/software/desi/users/sjbailey/redrock/py/redrock/zfind.py", line 202, in zfind
    local_targets, results = _rebalance_after_scan(targets, results)
Proc 0:   File "/global/common/software/desi/users/sjbailey/redrock/py/redrock/zfind.py", line 107, in _rebalance_after_scan
    lopsided_targets = targets.comm.gather(targets.local(), root=0)
Proc 0:   File "mpi4py/MPI/Comm.pyx", line 1578, in mpi4py.MPI.Comm.gather
Proc 0:   File "mpi4py/MPI/msgpickle.pxi", line 772, in mpi4py.MPI.PyMPI_gather
Proc 0:   File "mpi4py/MPI/msgpickle.pxi", line 147, in mpi4py.MPI.pickle_dump
Proc 0:   File "mpi4py/MPI/msgbuffer.pxi", line 50, in mpi4py.MPI.downcast
Proc 0: OverflowError: integer 2261775674 does not fit in 'int'

MPICH Notice [Rank 0] [job id 3839955.3] [Fri Dec  2 08:28:05 2022] [nid001228] - Abort(0) (rank 0 in comm 0): application called MPI_Abort(MPI_COMM_WORLD, 0) - process 0

aborting job:
application called MPI_Abort(MPI_COMM_WORLD, 0) - process 0
srun: error: nid001228: tasks 1-3: Exited with exit code 255
srun: launch/slurm: _step_signal: Terminating StepId=3839955.3
slurmstepd: error: *** STEP 3839955.3 ON nid001228 CANCELLED AT 2022-12-02T16:28:05 ***
srun: error: nid001228: task 0: Exited with exit code 255
srun: error: nid001228: tasks 4-63: Terminated
srun: Force Terminated StepId=3839955.3
```

And on other cases of the main branch on large files, like `/global/cfs/cdirs/desi/spectro/redux/himalayas/healpix/sv3/dark/260/26086/coadd-sv3-dark-26086.fits` with 1902 input targets (warning: this might get deleted and recreated during debugging re-runs):
```
--- Process 0 raised an exception ---
Proc 0: Traceback (most recent call last):
Proc 0:   File "/global/common/software/desi/perlmutter/desiconda/20220119-2.0.1/code/redrock/main/py/redrock/external/desi.py", line 787, in rrdesi
    scandata, zfit = zfind(targets, dtemplates, mpprocs,
Proc 0:   File "/global/common/software/desi/perlmutter/desiconda/20220119-2.0.1/code/redrock/main/py/redrock/zfind.py", line 230, in zfind
    local_targets, results = _rebalance_after_scan(targets, results)
Proc 0:   File "/global/common/software/desi/perlmutter/desiconda/20220119-2.0.1/code/redrock/main/py/redrock/zfind.py", line 115, in _rebalance_after_scan
    lopsided_targets = targets.comm.gather(targets.local(), root=0)
Proc 0:   File "mpi4py/MPI/Comm.pyx", line 1578, in mpi4py.MPI.Comm.gather
Proc 0:   File "mpi4py/MPI/msgpickle.pxi", line 773, in mpi4py.MPI.PyMPI_gather
Proc 0:   File "mpi4py/MPI/msgpickle.pxi", line 778, in mpi4py.MPI.PyMPI_gather
Proc 0:   File "mpi4py/MPI/msgpickle.pxi", line 191, in mpi4py.MPI.pickle_allocv
Proc 0:   File "mpi4py/MPI/msgpickle.pxi", line 182, in mpi4py.MPI.pickle_alloc
Proc 0: SystemError: Negative size passed to PyBytes_FromStringAndSize

MPICH Notice [Rank 0] [job id 3829709.0] [Thu Dec  1 14:14:12 2022] [nid002976] - Abort(0) (rank 0 in comm 0): application called MPI_Abort(MPI_COMM_WORLD, 0) - process 0

aborting job:
application called MPI_Abort(MPI_COMM_WORLD, 0) - process 0
```

I suspect that all of these cases are due to mpi4py using pickle to internally package objects for broadcast/gather operations, and it is probably exceeding the 2GB limit.

@craigwarner-ufastro @dmargala please take a look.  This is a blocking factor for us using GPU redrock for healpix redshifts.

If we can't readily fix redrock, alternatives are
  * use CPUs for healpix redshifts (at least on bigger files)
  * use larger nside = smaller files, but that's a little distasteful to have the densest region of the sky force the nside choice for the entire sky
  * restructure the healpix directory structure to dynamically adapt the nside to the density of the sky.  https://arxiv.org/pdf/1505.04632.pdf section 4 has a method to uniquely map dynamic nside+healpixnumber to an integer, but we haven't solved the inverse problem of how to make it easy for a user to know what file contains their RA,dec target if different files use different nside.
Any of these would be implemented on the desispec side, not the redrock side.

**Comment by craigwarner-ufastro** [comment]
Hi @sbailey @dmargala I have taken a look and can reproduce the error and I concur that the issue is in mpi4py using pickle internally and exceeding a 2 GB limit.

However on my GPU branch it runs with no problems with 4 GPU / 4 CPU (Total run time: 157.5s FYI) -- this of course does not do the rebalancing targets step.  So given that this should not be a problem once everything is merged, my instinct would be to not make any huge changes on the desispec side and rather you could either

- Use CPUs for bigger files (as mentioned)
- Use 4 GPU / 4 CPU -- though I suspect until my branch is merged this will be slower overall than 64 CPU or 4 GPU / 64 CPU
- We could also implement something like a --no-rebalance flag but that would essentially force the 4 GPU / 4 CPU scenario above so I like that less.

Additionally as this relates to our discussion on whether to remove rebalancing or not in the future, this should be considered as well as the timing results I sent on slack as to whether we want to just strip this out and do away with it once the GPU branch is merged with main.

**Comment by sbailey** [comment]
Thanks for looking into this @craigwarner-ufastro .  Currently the desispec pipeline wrapper to redrock [desispec.scripts.zproc](https://github.com/desihub/desispec/blob/2d42980f109f346a6c7f7d584549ce3908101141/py/desispec/scripts/zproc.py#L453) hands rrdesi the full MPI communicator and lets redrock decide what to do with it.  Would it be hard / problematic for redrock to decide to ignore the other ranks, or do you need the pipeline wrapper to create a sub-communicator with only 4 ranks first so that rrdesi only ever sees those 4?  zproc uses up to 30 ranks in parallel right now for other steps so the originating srun call can't just be -n 4.

**Comment by craigwarner-ufastro** [comment]
Hi @sbailey ah ok.  Let me look into that - this would seem to change the calculus on having a --no-rebalance option or something.  But it also may change the performance degradation with adding CPU cores that have no targets from a curiousity to something we care more about tracking down and solving if possible.

I will look into a couple of options and get back to you.

**Comment by sbailey** [comment]
Thanks @craigwarner-ufastro .  First priority is getting it into a state where `desispec.scripts.zproc.main` + `redrock.external.desi.rrdesi` doesn't crash on big inputs.  Second priority is any further optimizations available by using any leftover CPU ranks.

**Comment by craigwarner-ufastro** [comment]
@sbailey It turns out to be very trivial to just ignore the rebalancing:
`diff --git a/py/redrock/external/desi.py b/py/redrock/external/desi.py
index e6ad190..e2c2300 100644
--- a/py/redrock/external/desi.py
+++ b/py/redrock/external/desi.py
@@ -373,7 +373,8 @@ class DistTargetsDESI(DistTargets):
             self._proc_targets = distribute_work(comm_size,
                 self._keep_targets, weights=tweights)
         else:
-            self.is_lopsided = True
+            #self.is_lopsided = True
+            self.is_lopsided = False
             self._proc_targets = distribute_work(comm_size,
                 self._keep_targets, weights=tweights, capacities=self.capacities)
`

Essentially this will not affect CPU runs and on GPU runs where the targets have been initially balanced to only run on the GPU ranks, the rebalancing won't be done and only the GPU ranks will be used for zfind and Finalizing.  This does seem to still have the performance issues in zfind mentioned above but it will work without crashing.  Should I create a branch and PR?

**Comment by sbailey** [comment]
@craigwarner-ufastro I need to finish my data plenary talk for Monday morning so my attention isn't completely on this, but would it make sense to add this change to just your gpu branch for PR #211 ?

I'm thinking that for the current main branch, if the file is small enough to not blow memory we'd want to use GPU+CPU+rebalancing, and if it is too big then we're better off using only CPU than only GPU with no rebalancing.  But with your branch it would be handy to be able to hand it an arbitrary number of CPU ranks and have it ignore anything it doesn't need.

I'm still digesting all the tables, but back of the envelope 4 GPUs is 2x faster than 60 CPU cores, so in some future world it seems optimal to hand 2/3 of the work to the 4 GPUs and 1/3 of the work to 60 CPU cores, and finish in 2/3 of the time that it would have taken GPUs alone.

**Comment by craigwarner-ufastro** [comment]
@sbailey we could add a --no-rebalance command line option if you wanted for now that could be used with big files?  Because I do agree for smaller files where memory isn't an issue, we would like to do the rebalancing in the current main branch.  The other option would be to just use CPU only for big files for now (and I think this would be faster anyway than not rebalancing on the current branch).

For the future, 4 GPUs is more like 10x faster than 60 CPU cores for the coarse zscan, which is why the CPU only ranks get 0 targets - even giving a single target causes a slowdown when MPI overhead is included.  Moreover, the startup overhead of running with 64 cores takes more than double the overall clock time for running on 4 GPU / 4 CPU.  So while the fine zscan does run ever so slightly faster when rebalanced to 64 ranks (~2sec speed gain), even just the rebalancing and MPI overhead more than offsets this gain and the startup overhead dwarfs it.

Once the GPU branch is merged, it would be optimal to use 4 GPU / 4 CPU (which avoids pickle memory issues of course).  I agree it would be nice to be able to hand it an arbitrary number of CPU ranks and have it ignore the others -- so it would make sense to include that in my massive PR -- and hopefully track down why having non-GPU ranks seems to slow down the fine redshift scan.

**Comment by sbailey** [comment]
fixed by PR #211.  Closing.
